src/ess/powder/nexus.py

```diff
@@ -0,0 +1,251 @@
+
+
+def get_sample_position(
+    raw_sample: RawSample[RunType],
+) -> SamplePosition[RunType]:
+    return SamplePosition[RunType](raw_sample["position"])
+
+
+def get_detector_data(
+    detector: NeXusDetector[RunType],
+    bank_sizes: DetectorBankSizes | None = None,
+) -> RawDetector[RunType]:
+    da = nexus.extract_detector_data(detector)
+    if (sizes := (bank_sizes or {}).get(detector['detector_name'])) is not None:
+        da = da.fold(dim="detector_number", sizes=sizes)
+    return RawDetector[RunType](da)
+
+
+def get_monitor_data(
+    monitor: NeXusMonitor[RunType, MonitorType],
+    source_position: SourcePosition[RunType],
+) -> RawMonitor[RunType, MonitorType]:
+    return RawMonitor[RunType, MonitorType](
+        nexus.extract_monitor_data(monitor).assign_coords(
+            position=monitor['position'], source_position=source_position
+        )
+    )
+
+
+def assemble_detector_data(
+    detector: RawDetector[RunType],
+    event_data: DetectorEventData[RunType],
+    source_position: SourcePosition[RunType],
+    sample_position: SamplePosition[RunType],
+) -> ReducibleDetectorData[RunType]:
+    """
+    Assemble a detector data object with source and sample positions and event data.
+    Also adds variances to the event data if they are missing.
+    """
+    grouped = nexus.group_event_data(
+        event_data=event_data, detector_number=detector.coords['detector_number']
+    )
+    detector.data = grouped.data
+    return ReducibleDetectorData[RunType](
+        _add_variances(da=detector).assign_coords(
+            source_position=source_position, sample_position=sample_position
+        )
+    )
+
+
+def assemble_monitor_data(
+    monitor_data: RawMonitor[RunType, MonitorType],
+    event_data: MonitorEventData[RunType, MonitorType],
+) -> RawMonitorData[RunType, MonitorType]:
+    meta = monitor_data.drop_coords('event_time_zero')
+    da = event_data.assign_coords(meta.coords).assign_masks(meta.masks)
+    return RawMonitorData[RunType, MonitorType](_add_variances(da=da))
+
+
+def _skip(
+    _: str, obj: snx.Field | snx.Group, classes: tuple[snx.NXobject, ...]
+) -> bool:
+    return isinstance(obj, snx.Group) and (obj.nx_class in classes)
+
+
+class FilteredDetector(snx.NXdetector):
+    def __init__(
+        self, attrs: dict[str, Any], children: dict[str, snx.Field | snx.Group]
+    ):
+        children = {
+            name: child
+            for name, child in children.items()
+            if not _skip(name, child, classes=(snx.NXoff_geometry,))
+        }
+        super().__init__(attrs=attrs, children=children)
+
+
+class NXmonitor_no_events(snx.NXmonitor):
+    def __init__(
+        self, attrs: dict[str, Any], children: dict[str, snx.Field | snx.Group]
+    ):
+        children = {
+            name: child
+            for name, child in children.items()
+            if not _skip(name, child, classes=(snx.NXevent_data,))
+        }
+        super().__init__(attrs=attrs, children=children)
```

Comment:
To avoid duplication, you could use something like this:
```suggestion
def filtered_nx_class(base_class, exclude):
    class Filtered(base_class):
        def __init__(
                self, attrs: dict[str, Any], children: dict[str, snx.Field | snx.Group]
        ):
            children = {
                name: child
                for name, child in children.items()
                if not _skip(name, child, classes=exclude)
            }
            super().__init__(attrs=attrs, children=children)
    return Filtered
```